crates/bevy_ecs/src/system/query.rs

```diff
@@ -50,330 +57,424 @@ use core::{
+/// fn immutable_query(query: Query<&ComponentA>) {
+///     // ...
+/// }
+///
+/// // ...or by a mutable reference.
+/// fn mutable_query(query: Query<&mut ComponentA>) {
+///     // ...
+/// }
+/// #
+/// # bevy_ecs::system::assert_is_system(immutable_query);
+/// # bevy_ecs::system::assert_is_system(mutable_query);
+/// ```
+///
+/// Note that components need to be behind a reference (`&` or `&mut`), or the query will not compile:
+///
+/// ```compile_fail,E0277
+/// # use bevy_ecs::prelude::*;
+/// #
+/// # #[derive(Component)]
+/// # struct ComponentA;
+/// #
+/// // This needs to be `&ComponentA` or `&mut ComponentA` in order to compile.
+/// fn invalid_query(query: Query<ComponentA>) {
+///     // ...
+/// }
 /// ```
 ///
 /// ## Query filtering
 ///
-/// Setting the query filter type parameter will ensure that each query item satisfies the given condition.
+/// Setting the query filter type parameter will ensure that each query item satisfies the given condition:
 ///
 /// ```
 /// # use bevy_ecs::prelude::*;
+/// #
 /// # #[derive(Component)]
 /// # struct ComponentA;
+/// #
 /// # #[derive(Component)]
 /// # struct ComponentB;
-/// # fn system(
-/// // Just `ComponentA` data will be accessed, but only for entities that also contain
-/// // `ComponentB`.
-/// query: Query<&ComponentA, With<ComponentB>>
-/// # ) {}
-/// # bevy_ecs::system::assert_is_system(system);
+/// #
+/// // `ComponentA` data will be accessed, but only for entities that also contain `ComponentB`.
+/// fn filtered_query(query: Query<&ComponentA, With<ComponentB>>) {
+///     // ...
+/// }
+/// #
+/// # bevy_ecs::system::assert_is_system(filtered_query);
 /// ```
 ///
+/// Note that the filter is `With<ComponentB>`, not `With<&ComponentB>`. Unlike query data, `With`
+/// does require components to be behind a reference.
+///
 /// ## `QueryData` or `QueryFilter` tuples
 ///
-/// Using tuples, each `Query` type parameter can contain multiple elements.
+/// Using [`tuple`]s, each `Query` type parameter can contain multiple elements.
+///
+/// In the following example two components are accessed simultaneously, and the query items are
+/// filtered on two conditions:
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// #
+/// # #[derive(Component)]
+/// # struct ComponentA;
+/// #
+/// # #[derive(Component)]
+/// # struct ComponentB;
+/// #
+/// # #[derive(Component)]
+/// # struct ComponentC;
+/// #
+/// # #[derive(Component)]
+/// # struct ComponentD;
+/// #
+/// fn complex_query(
+///     query: Query<(&mut ComponentA, &ComponentB), (With<ComponentC>, Without<ComponentD>)>
+/// ) {
+///     // ...
+/// }
+/// #
+/// # bevy_ecs::system::assert_is_system(complex_query);
+/// ```
 ///
-/// In the following example, two components are accessed simultaneously, and the query items are filtered on two conditions.
+/// Note that this currently only works on tuples with 15 or fewer items. You may nest tuples,
+/// however, to get around this limit:
```

Comment:
Nit: That however sounds odd to me. Maybe just get rid of it. 

"You can nest tuples to get around this limit if needed" or something